### PR TITLE
GAX: LRO HTTP headers

### PR DESCRIPTION
**Pull request by lukesneeringer**
This pull requests adds the HTTP headers implementation to `OperationsClient` so that they are sent on LRO calls. It also moves some of the functionality in `api_callable` into `utils.metrics` for reusability.

### CODE REVIEW

File: google/gax/utils/metrics.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        answer[k] = v
+
+    # Add the GAX and GRPC headers to our metrics.
+    # These come after what may have been passed in (generally the GAPIC
+    # library).
+    answer['gax'] = gax.__version__
+    # pylint: disable=no-member
+    answer['grpc'] = pkg_resources.get_distribution('grpcio').version
+    # pylint: enable=no-member
+
```
Comment (geigerj) [inline google/gax/utils/metrics.py]: No special treatment for a/b keys?

File: google/gapic/longrunning/operations_client.py [hunk 2]
```diff
@@ -122,10 +125,17 @@ def __init__(self,
...
-        metadata = [('x-goog-api-client', goog_api_client)]
+
+        # Construct the correct X-Goog-API-Client header, as a string.
+        # This ends up being in the following format:
+        #   [('x-goog-api-client', 'gax/x.y.z grpc/x.y.z')]
+        #
+        # This is the format that the underlying GRPC layer expects.
+        metadata = [(
+            'x-goog-api-client',
+            metrics.stringify(metrics.fill(metrics_headers)),
```
Comment (geigerj) [inline google/gapic/longrunning/operations_client.py]: I don't see this in toolkit -- would very strongly prefer that this file not be hand-edited. Why does this need to be different from the normal GAPICs?

File: google/gax/utils/metrics.py (hunk 1 above)
Comment (lukesneeringer) [inline google/gax/utils/metrics.py]: Nope, we nixed that from the spec at Hong's request. (It is too different in purpose from everything else.)

File: google/gapic/longrunning/operations_client.py (hunk 2 above)
Comment (lukesneeringer) [inline google/gapic/longrunning/operations_client.py]: Is this file _generated_? Why is it in GAX if it is generated?

Comment (geigerj) [inline google/gapic/longrunning/operations_client.py]: > Is this file generated?

Yes, this directory is ordinary GAPIC output.

> Why is it in GAX if it is generated?

@bjwatson, could you provide the context here?

File: tests/test_utils_metrics.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        # Assert that the headers are in the correct order.
+        self.assertEqual(
+            [k for k in headers.keys()],
+            ['gl-python', 'gapic', 'gax', 'grpc'],
+        )
+
+
+class TestStringify(unittest.TestCase):
+    def test_no_argument(self):
+        s = metrics.stringify()
```
Comment (geigerj) [inline tests/test_utils_metrics.py]: Please don't use a single-character variable name here and below: Google Python style guide permits them only as [counters and iterators](https://google.github.io/styleguide/pyguide.html#Naming).

Something more descriptive (like `empty_metrics`?) would be nice.

File: tests/test_utils_metrics.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+        # Assert that each section of the string is present, but
+        # ignore ordering.
+        self.assertIn('gl-python/%s' % platform.python_version(), s)
+        self.assertIn('gapic/1.0.0', s)
+        self.assertIn('gax/%s' % gax.__version__, s)
+        self.assertIn('grpc/%s' % GRPC_VERSION, s)
+
+    def test_with_ordered_argument(self):
+        mh = collections.OrderedDict()
```
Comment (geigerj) [inline tests/test_utils_metrics.py]: similarly, would prefer `metrics_headers` to `mh`.

File: tests/test_utils_metrics.py (hunk 3 above)
Comment (lukesneeringer) [inline tests/test_utils_metrics.py]: Changed.

File: tests/test_utils_metrics.py (hunk 4 above)
Comment (lukesneeringer) [inline tests/test_utils_metrics.py]: Changed.

**Pull request merged**